File: library-ui/src/app/interfaces/book-details.ts
import {BasicCollection} from './basic-collection';
import {LibraryBook} from './library-book';
import {Book} from './book';

export const BOOK_DETAIL_TYPES = {
  GLOBAL: 'GLOBAL',
  LIBRARY: 'LIBRARY',
} as const;

export type BookDetailType = typeof BOOK_DETAIL_TYPES[keyof typeof BOOK_DETAIL_TYPES];

export interface BaseBookDetails {
  type: BookDetailType;
  averageRating: number;
  ratingsNumber: number;
}

export interface GlobalBookDetails extends BaseBookDetails {
  type: typeof BOOK_DETAIL_TYPES.GLOBAL;
  book: Book;
}

export interface LibraryBookDetails extends BaseBookDetails {
  type: typeof BOOK_DETAIL_TYPES.LIBRARY;
  libraryBook: LibraryBook;
  collections: BasicCollection[];
}

export type BookDetails = GlobalBookDetails | LibraryBookDetails;

File: library-app/src/main/java/org/example/library/book/dto/BookDetails.java
package org.example.library.book.dto;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = GlobalBookDetails.class, name = "GLOBAL"),
        @JsonSubTypes.Type(value = LibraryBookDetails.class, name = "LIBRARY")
})
@Getter
@RequiredArgsConstructor
public abstract class BookDetails {

    private final double averageRating;
    private final long ratingsNumber;

}
